File: habitune-frontend/src/types/iteration2.ts
export type SpeciesGroup = 'native_bee' | 'butterfly' | 'small_bird'

export type DataAvailability = 'idle' | 'loading' | 'available' | 'empty' | 'unavailable' | 'not_found' | 'error'

export type GeoJsonGeometry = {
  type: 'Point' | 'LineString' | 'Polygon' | 'MultiLineString' | 'MultiPolygon'
  coordinates: unknown[]
}

export interface LocationContext {
  contextId?: string | null
  precinctId?: string | null
  precinctName?: string | null
  streetId?: string | null
  streetKey?: string | null
  sourceStreetId?: string | null
  streetName?: string | null
  resolutionMethod?: string | null
  matchedAddress?: string | null
  distanceToAddressM?: number | null
  label?: string | null
  latitude: number
  longitude: number
  geometry?: GeoJsonGeometry | null
}

export interface HabitatFeature {
  id: string
  kind: 'tree' | 'canopy' | 'garden' | 'vegetation' | 'other'
  label: string
  evidenceSource?: string | null
  geometry?: GeoJsonGeometry | null
}

export interface NearbyObservation {
  id: string
  speciesGroup: SpeciesGroup
  scientificName?: string | null
  commonName?: string | null
  occurrenceCount?: number | null
  observedAt?: string | null
  geometry?: GeoJsonGeometry | null
}

export interface ConnectivityGap {
  id: string
  label: string
  explanation: string
  severity?: 'potential' | 'unknown'
  geometry?: GeoJsonGeometry | null
}

export interface PotentialConnectivity {
  id: string
  speciesGroup: SpeciesGroup
  status: 'not_available' | 'provisional'
  supportLevel?: 'strong' | 'moderate' | 'potential_gap' | null
  explanation?: string | null
  geometry?: GeoJsonGeometry | null
  gaps: ConnectivityGap[]
}

export interface StreetPollinatorSupport {
  streetId: string
  streetKey?: string | null
  streetName: string
  precinctId?: string | null
  nearbyCanopyAreaM2?: number | null
  geometry?: GeoJsonGeometry | null
  centroid?: { latitude: number; longitude: number } | null
  sourceStreetId?: string | null
  addressCount?: number | null
  plantedTreeCount?: number | null
  plantedTreeSpeciesCount?: number | null
  gardenPlantRowCount?: number | null
  gardenPlantSpeciesCount?: number | null
  plantSpeciesCount?: number | null
  pollinatorFloweringPlantSpeciesCount?: number | null
  nearbyCanopyPolygonCount?: number | null
}

export interface ConnectivityData {
  context: LocationContext | null
  supportRecords: StreetPollinatorSupport[]
  connectivity: PotentialConnectivity[]
  connectivityStatus: 'not_modelled'
}

export interface StreetEcosystemContext {
  streetId: string
  streetKey?: string | null
  sourceStreetId?: string | null
  precinctId?: string | null
  streetName?: string | null
  centroid?: { latitude: number; longitude: number } | null
  evidence?: StreetEcosystemEvidence | null
  location?: LocationContext | null
  speciesGroup?: SpeciesGroup | null
  habitatFeatures: HabitatFeature[]
  nearbyObservations: NearbyObservation[]
  potentialConnectivity?: PotentialConnectivity | null
  support?: StreetPollinatorSupport | null
  status: 'not_available' | 'available'
  message?: string | null
}

export interface StreetEcosystemEvidence {
  addressCount: number | null
  plantedTreeCount: number | null
  plantedTreeSpeciesCount: number | null
  gardenPlantRowCount: number | null
  gardenPlantSpeciesCount: number | null
  plantSpeciesCount: number | null
  pollinatorFloweringPlantSpeciesCount: number | null
  nearbyCanopyAreaM2: number | null
  nearbyCanopyPolygonCount: number | null
}

export interface ApiEnvelope<T> {
  data: T
  meta?: {
    source?: string
    status?: string
    methodVersion?: string | null
    geometryStatus?: string
    resolutionSourceFile?: string | null
    streetSourceFile?: string | null
    sourceFile?: string | null
    assignmentMethod?: string | null
    maximumAssignmentDistanceM?: number | null
    canopyMetricNote?: string | null
    connectivityStatus?: 'not_modelled'
  }
}

File: habitune-backend/shared/contracts/iteration2.py
"""Iteration 2 API contracts with deliberately optional geometry."""

from __future__ import annotations

from typing import Any, Literal, NotRequired, TypedDict


SpeciesGroup = Literal["native_bee", "butterfly", "small_bird"]


class Geometry(TypedDict):
    type: Literal["Point", "LineString", "Polygon", "MultiLineString", "MultiPolygon"]
    coordinates: list


class LocationContext(TypedDict):
    contextId: NotRequired[str | None]
    precinctId: NotRequired[str | None]
    streetId: NotRequired[str | None]
    streetKey: NotRequired[str | None]
    sourceStreetId: NotRequired[str | None]
    precinctName: NotRequired[str | None]
    streetName: NotRequired[str | None]
    resolutionMethod: NotRequired[str | None]
    matchedAddress: NotRequired[str | None]
    distanceToAddressM: NotRequired[float | None]
    label: NotRequired[str | None]
    latitude: float
    longitude: float
    geometry: NotRequired[Geometry | None]


class HabitatFeature(TypedDict):
    id: str
    kind: Literal["tree", "canopy", "garden", "vegetation", "other"]
    label: str
    evidenceSource: NotRequired[str | None]
    geometry: NotRequired[Geometry | None]


class NearbyObservation(TypedDict):
    id: str
    speciesGroup: SpeciesGroup
    scientificName: NotRequired[str | None]
    commonName: NotRequired[str | None]
    occurrenceCount: NotRequired[int | None]
    observedAt: NotRequired[str | None]
    geometry: NotRequired[Geometry | None]


class ConnectivityGap(TypedDict):
    id: str
    label: str
    explanation: str
    severity: NotRequired[Literal["potential", "unknown"]]
    geometry: NotRequired[Geometry | None]


class PotentialConnectivity(TypedDict):
    id: str
    speciesGroup: SpeciesGroup
    status: Literal["not_available", "provisional"]
    supportLevel: NotRequired[Literal["strong", "moderate", "potential_gap"] | None]
    explanation: NotRequired[str | None]
    geometry: NotRequired[Geometry | None]
    gaps: list[ConnectivityGap]


class StreetPollinatorSupport(TypedDict):
    streetId: str
    streetKey: NotRequired[str]
    streetName: str
    precinctId: NotRequired[str | None]
    nearbyCanopyAreaM2: NotRequired[float | None]
    geometry: NotRequired[Geometry | None]
    centroid: NotRequired[dict[str, float] | None]
    sourceStreetId: NotRequired[str | None]
    addressCount: NotRequired[int | None]
    plantedTreeCount: NotRequired[int | None]
    plantedTreeSpeciesCount: NotRequired[int | None]
    gardenPlantRowCount: NotRequired[int | None]
    gardenPlantSpeciesCount: NotRequired[int | None]
    plantSpeciesCount: NotRequired[int | None]
    pollinatorFloweringPlantSpeciesCount: NotRequired[int | None]
    nearbyCanopyAreaM2: NotRequired[float | None]
    nearbyCanopyPolygonCount: NotRequired[int | None]


class ConnectivityData(TypedDict):
    context: LocationContext | None
    supportRecords: list[StreetPollinatorSupport]
    connectivity: list[PotentialConnectivity]
    connectivityStatus: Literal["not_modelled"]


class StreetEcosystemContext(TypedDict):
    streetId: str
    streetKey: NotRequired[str]
    sourceStreetId: NotRequired[str]
    precinctId: NotRequired[str]
    streetName: NotRequired[str | None]
    centroid: NotRequired[dict[str, float] | None]
    evidence: NotRequired[StreetEcosystemEvidence]
    location: NotRequired[LocationContext | None]
    speciesGroup: NotRequired[SpeciesGroup | None]
    habitatFeatures: list[HabitatFeature]
    nearbyObservations: list[NearbyObservation]
    potentialConnectivity: NotRequired[PotentialConnectivity | None]
    status: Literal["not_available", "available"]
    message: NotRequired[str | None]


class StreetEcosystemEvidence(TypedDict):
    addressCount: int | None
    plantedTreeCount: int | None
    plantedTreeSpeciesCount: int | None
    gardenPlantRowCount: int | None
    gardenPlantSpeciesCount: int | None
    plantSpeciesCount: int | None
    pollinatorFloweringPlantSpeciesCount: int | None
    nearbyCanopyAreaM2: float | None
    nearbyCanopyPolygonCount: int | None


def api_meta(*, source="not_integrated", status="not_available", geometry_status="not_available", method_version=None, **metadata):
    """Build the shared metadata shape used by Iteration 2 endpoints."""
    return {
        "source": source,
        "status": status,
        "methodVersion": method_version,
        "geometryStatus": geometry_status,
        **metadata,
    }


def api_envelope(data: Any, *, source="not_integrated", status="not_available", geometry_status="not_available", method_version=None, **metadata):
    """Construct an API envelope through the contract module to prevent shape drift."""
    return {"data": data, "meta": api_meta(source=source, status=status, geometry_status=geometry_status, method_version=method_version, **metadata)}


def connectivity_data(*, context=None, support_records=None, connectivity=None) -> ConnectivityData:
    return {
        "context": context,
        "supportRecords": list(support_records or []),
        "connectivity": list(connectivity or []),
        "connectivityStatus": "not_modelled",
    }
